token-metadata/program/tests/verify_sized_collection_item.rs

```diff
@@ -1284,3 +1284,336 @@ mod verify_sized_collection_item {
+        .await
+        .unwrap();
+
+    let name = "Test".to_string();
+    let symbol = "TST".to_string();
+    let uri = "uri".to_string();
+    let test_metadata = Metadata::new();
+    test_metadata
+        .create_v2(
+            &mut context,
+            name,
+            symbol,
+            uri,
+            None,
+            10,
+            false,
+            None,
+            Some(Collection {
+                key: test_collection.mint.pubkey(),
+                verified: false,
+            }),
+            None,
+        )
+        .await
+        .unwrap();
+
+    let metadata = test_metadata.get_data(&mut context).await;
+    assert_eq!(
+        metadata.collection.to_owned().unwrap().key,
+        test_collection.mint.pubkey()
+    );
+    assert!(!metadata.collection.unwrap().verified);
+
+    let kpbytes = &context.payer;
+    let kp = Keypair::from_bytes(&kpbytes.to_bytes()).unwrap();
+
+    test_metadata
+        .verify_sized_collection_item(
+            &mut context,
+            test_collection.pubkey,
+            &kp,
+            test_collection.mint.pubkey(),
+            collection_master_edition_account.pubkey,
+            None,
+        )
+        .await
+        .unwrap();
+
+    let metadata_after = test_metadata.get_data(&mut context).await;
+    assert_eq!(
+        metadata_after.collection.to_owned().unwrap().key,
+        test_collection.mint.pubkey()
+    );
+    assert!(metadata_after.collection.unwrap().verified);
+
+    let collection_md = test_collection.get_data(&mut context).await;
+    let size = if let Some(details) = collection_md.collection_details {
+        match details {
+            CollectionDetails::V1 { size } => size,
+        }
+    } else {
+        panic!("Expected CollectionDetails::V1");
+    };
+    assert_eq!(size, 1);
+
+    // Allow time to pass so our tx isn't rejected for being 'already processed'.
+    context.warp_to_slot(100).unwrap();
+
+    let error = test_metadata
+        .verify_sized_collection_item(
+            &mut context,
+            test_collection.pubkey,
+            &kp,
+            test_collection.mint.pubkey(),
+            collection_master_edition_account.pubkey,
+            None,
+        )
+        .await
+        .unwrap_err();
+
+    let metadata_after = test_metadata.get_data(&mut context).await;
+    assert_eq!(
+        metadata_after.collection.to_owned().unwrap().key,
+        test_collection.mint.pubkey()
+    );
+    assert!(metadata_after.collection.unwrap().verified);
+    assert_custom_error!(error, MetadataError::AlreadyVerified);
+
+    let collection_md = test_collection.get_data(&mut context).await;
+    let size = if let Some(details) = collection_md.collection_details {
+        match details {
+            CollectionDetails::V1 { size } => size,
+        }
+    } else {
+        panic!("Expected CollectionDetails::V1");
+    };
+    assert_eq!(size, 1);
+}
+
+#[tokio::test]
+async fn fail_unverify_already_unverified() {
```

Comment:
This might be one portion of what Austin is already saying, but I would like to add to this comment specifically:

I looked briefly through the existing `verify_sized_collection_item` tests, and I am seeing all the good checks to be defensive against missing or incorrect data or signers, etc.  and I'm seeing success case with collection size of 1.

But to me a piece that we should specifically test is that `CollectionDetailsV1::size` is working properly, i.e., incrementing and decrementing in general, including even decrementing past 0 case (probably not incrementing past `u64::max` unless you can hardcode it somehow).

I could just be missing where these size checks are occurring.

Reply:
Done.

Reply:
Talked offline with @samuelvanderwaal and he showed me that the `size_tracking` module in `collection_sizes.rs` is what I was looking for.  Comment resolved.